File: cores/csharp/src/Mappers/Cnrom.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// Mapper 3 (CNROM). Port of src/mappers/cnrom.rs.
public sealed class Cnrom : Mapper
{
    private const uint PrgBankSize = 16384;
    private const uint ChrBankSize = 8192;

    private byte[] _prgRom;
    private uint _prgSize;
    private byte[] _chr;
    private uint _chrSize;
    private bool _chrIsRam;
    private Mirroring _mirroring;
    private bool _hasBattery;
    private byte _chrBank;

    public Cnrom(byte[] prg, uint prgSize, byte[] chr, uint chrSize, Mirroring mirroring, bool hasBattery)
    {
        MapperNum = 3;
        _prgSize = prgSize;
        _prgRom = new byte[prgSize];
        if (prgSize > 0) System.Array.Copy(prg, _prgRom, (int)prgSize);
        if (chrSize == 0)
        {
            _chrSize = ChrBankSize;
            _chr = new byte[ChrBankSize];
            _chrIsRam = true;
        }
        else
        {
            _chrSize = chrSize;
            _chr = new byte[chrSize];
            System.Array.Copy(chr, _chr, (int)chrSize);
            _chrIsRam = false;
        }
        _mirroring = mirroring;
        _hasBattery = hasBattery;
        _chrBank = 0;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private uint ChrBankCount() => _chrSize / ChrBankSize > 0 ? _chrSize / ChrBankSize : 1;
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private uint PrgIndex(ushort addr)
    {
        uint local = (uint)(addr - 0x8000);
        uint bank = _prgSize;
        if (bank == 0) return 0;
        return local % bank;
    }

    public override byte ReadPrg(ushort addr)
    {
        if (addr < 0x8000) return 0x00;
        uint idx = PrgIndex(addr);
        if (idx >= _prgSize) return 0x00;
        return _prgRom[idx];
    }

    public override void WritePrg(ushort addr, byte value)
    {
        if (addr < 0x8000) return;
        _chrBank = value;
    }

    public override byte ReadChr(ushort addr)
    {
        uint count = ChrBankCount();
        uint bank = _chrBank % count;
        uint idx = bank * ChrBankSize + ((uint)addr & (ChrBankSize - 1));
        if (idx >= _chrSize) return 0x00;
        return _chr[idx];
    }

    public override void WriteChr(ushort addr, byte value)
    {
        if (!_chrIsRam) return;
        uint count = ChrBankCount();
        uint bank = _chrBank % count;
        uint idx = bank * ChrBankSize + ((uint)addr & (ChrBankSize - 1));
        if (idx < _chrSize) _chr[idx] = value;
    }

    public override Mirroring MirrorMode() => _mirroring;
    public override bool ChrIsRam() => _chrIsRam;
    public override bool HasBattery() => _hasBattery;

    public override uint SaveState(byte[] buf)
    {
        uint total = 1 + (_chrIsRam ? _chrSize : 0);
        if (buf == null) return total;
        int p = 0;
        buf[p++] = _chrBank;
        if (_chrIsRam && _chrSize > 0)
        {
            System.Array.Copy(_chr, 0, buf, p, (int)_chrSize);
            p += (int)_chrSize;
        }
        return (uint)p;
    }

    public override bool LoadState(byte[] buf, uint len)
    {
        uint need = 1 + (_chrIsRam ? _chrSize : 0);
        if (len < need) return false;
        int p = 0;
        _chrBank = buf[p++];
        if (_chrIsRam && _chrSize > 0)
        {
            System.Array.Copy(buf, p, _chr, 0, (int)_chrSize);
            p += (int)_chrSize;
        }
        return true;
    }
}

File: cores/typescript/src/mappers/cnrom.ts
// Mapper 3 (CNROM). Port of cores/c/src/mappers/cnrom.c.

import { Mapper, Mirroring } from '../mapper';

export class Cnrom extends Mapper {
    private static readonly PrgBankSize = 16384;
    private static readonly ChrBankSize = 8192;

    private _prgRom: Uint8Array;
    private _prgSize: number;
    private _chr: Uint8Array;
    private _chrSize: number;
    private _chrIsRam: boolean;
    private _mirroring: number;
    private _hasBattery: boolean;
    private _chrBank: number;

    constructor(prg: Uint8Array, prgSize: number, chr: Uint8Array | null, chrSize: number, mirroring: number, hasBattery: boolean) {
        super();
        this.mapperNum = 3;
        this._prgSize = prgSize;
        this._prgRom = new Uint8Array(prgSize);
        if (prgSize > 0) this._prgRom.set(prg.subarray(0, prgSize));
        if (chrSize === 0) {
            this._chrSize = Cnrom.ChrBankSize;
            this._chr = new Uint8Array(Cnrom.ChrBankSize);
            this._chrIsRam = true;
        } else {
            this._chrSize = chrSize;
            this._chr = new Uint8Array(chrSize);
            this._chr.set(chr!.subarray(0, chrSize));
            this._chrIsRam = false;
        }
        this._mirroring = mirroring;
        this._hasBattery = hasBattery;
        this._chrBank = 0;
    }

    private chrBankCount(): number {
        const c = Math.floor(this._chrSize / Cnrom.ChrBankSize);
        return c > 0 ? c : 1;
    }

    private prgIndex(addr: number): number {
        const local = addr - 0x8000;
        const bank = this._prgSize;
        if (bank === 0) return 0;
        return local % bank;
    }

    readPrg(addr: number): number {
        if (addr < 0x8000) return 0x00;
        const idx = this.prgIndex(addr);
        if (idx >= this._prgSize) return 0x00;
        return this._prgRom[idx];
    }

    writePrg(addr: number, value: number): void {
        if (addr < 0x8000) return;
        this._chrBank = value;
    }

    readChr(addr: number): number {
        const count = this.chrBankCount();
        const bank = this._chrBank % count;
        const idx = bank * Cnrom.ChrBankSize + (addr & (Cnrom.ChrBankSize - 1));
        if (idx >= this._chrSize) return 0x00;
        return this._chr[idx];
    }

    writeChr(addr: number, value: number): void {
        if (!this._chrIsRam) return;
        const count = this.chrBankCount();
        const bank = this._chrBank % count;
        const idx = bank * Cnrom.ChrBankSize + (addr & (Cnrom.ChrBankSize - 1));
        if (idx < this._chrSize) this._chr[idx] = value;
    }

    mirrorMode(): number { return this._mirroring; }
    chrIsRam(): boolean { return this._chrIsRam; }
    hasBattery(): boolean { return this._hasBattery; }

    saveState(buf: Uint8Array | null): number {
        const total = 1 + (this._chrIsRam ? this._chrSize : 0);
        if (buf === null) return total;
        let p = 0;
        buf[p++] = this._chrBank;
        if (this._chrIsRam && this._chrSize > 0) {
            buf.set(this._chr.subarray(0, this._chrSize), p);
            p += this._chrSize;
        }
        return p;
    }

    loadState(buf: Uint8Array, len: number): boolean {
        const need = 1 + (this._chrIsRam ? this._chrSize : 0);
        if (len < need) return false;
        let p = 0;
        this._chrBank = buf[p++];
        if (this._chrIsRam && this._chrSize > 0) {
            this._chr.set(buf.subarray(p, p + this._chrSize));
            p += this._chrSize;
        }
        return true;
    }
}
